### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by nzpr**
### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: v2/src/main/scala/coop/rchain/v2/casper/data/CasperScope.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * @param latestMessages      Upper boundary, latest messages known. Define the whole [[CasperScope]].
+ * @param finalizationFringe  Bottom boundary, set of finalized messages for all known senders.
+ * @param conflictScope       The body of the scope - all messages above [[finalizationFringe]].
+ */
+final case class CasperScope[M, S](
+    latestMessages: LatestMessages[M, S],
+    finalizationFringe: FinalizationFringe[M, S],
+    conflictScope: ConflictScope[M]
+) { require(finalizationFringe.v.keySet == latestMessages.v.keySet) }
```
Comment (zsluedem) [inline v2/src/main/scala/coop/rchain/v2/casper/data/CasperScope.scala:17]: `Casper scope is valid only if finalization fringe is complete: finalized messages are found for all known senders.`

What does `valid` mean in the above sentence?

Can I say "For a particular block, there is always a fixed value `CasperScope` corresponding to this particular block" is a correct property for a block?

File: v2/src/main/scala/coop/rchain/v2/casper/data/FinalizationFringe.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ */
+final case class FinalizationFringe[M, S](v: Map[S, M])
+
+object FinalizationFringe {
+
+  /**
+   * Find messages that are finalized in next compared to prev.
+   */
+  def diff[M, S](next: FinalizationFringe[M, S], prev: FinalizationFringe[M, S]): Set[M] = ???
+}
```
Comment (zsluedem) [inline v2/src/main/scala/coop/rchain/v2/casper/data/FinalizationFringe.scala:17]: Sorry that I am still kind of unkown on the term `FinalizedFringe`. Is it possible to explain it in an example?

File: v2/src/main/scala/coop/rchain/v2/casper/Casper.scala [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+ * @param dg            Dependency graph exposing justification structure.
+ * @param so            Safety oracle, defining fault tolerances.
+ * @param maxComplexity Maximum number of traversals through dependency graph.
+ * @tparam M            Type of a message.
+ * @tparam S            Type of a sender.
+ */
+final case class Casper[M, S](
+    dg: DependencyGraph[M, S],
+    so: SafetyOracle[M, S],
+    maxComplexity: Long = Long.MaxValue
```
Comment (zsluedem) [inline v2/src/main/scala/coop/rchain/v2/casper/Casper.scala:19]: It is kind of wired to have a limit here.

File: v2/src/main/scala/coop/rchain/v2/casper/data/CasperScope.scala (hunk 1 above)
Comment (nzpr) [reply v2/src/main/scala/coop/rchain/v2/casper/data/CasperScope.scala:17]: Yes, you are right, for each block there is Casper scope associated. When you replay - you reconstruct the scope. To be able to create a block, proposer first have to find the scope, as well.

`Valid` means that if there is no finalized message for all senders in a fringe - Casper scope cannot be created. So if you will try to create such an instance - you will hit an error. I need to add here error msg actually.

File: v2/src/main/scala/coop/rchain/v2/casper/data/FinalizationFringe.scala (hunk 2 above)
Comment (nzpr) [reply v2/src/main/scala/coop/rchain/v2/casper/data/FinalizationFringe.scala:17]: `Finalization fringe` is a set of messages: highest finalized message for each known sender. Known senders are defined by latest messages, which you use to run Casper.

Screenshot 2021-11-02 at 12 33 08

Here you can see grey blocks: they are finalized. Collectively they are finalization fringe. Everything below them is finalized (conflict resolution has been done and acceptances and rejects are final).
 
(Also you can see there on extra grey block at the bottom - this is highest common message for messages of the fringe, it is required to merge fringe and get finalized state).

File: v2/src/main/scala/coop/rchain/v2/casper/Casper.scala (hunk 3 above)
Comment (nzpr) [reply v2/src/main/scala/coop/rchain/v2/casper/Casper.scala:19]: I'm trying to push a narrative that Casper is an algorithm, and we can limit its complexity if we want. So we cannot end up with Casper running for hours without knowing about that. We put a constraint on amount of resources that are allocated for Casper to search for a scope. If DAG is very diverged, Casper might not find a scope. Then we might increase complexity and try again, having the same DAG. Or just fail and throw some error.

**Comment by nzpr** [comment]
Closing as outdated

**Pull request closed without merging**